content/blog/2022-05-19-CoC/index.md

```diff
@@ -0,0 +1,298 @@
+project, we only implemented CoC.
+
+
+## Design and Implementation
+
+### Typing rules
+
+Unlike most other languages, CoC eliminates the distinction between types and terms. 
+We describe the CoC syntax below. We use $ M, N, ... $ for terms in general
+and $ x, y, z $ for variables.
+<p align="center">
+    $ M ::= x \; | \; (\lambda x:M) N \; | \; (M N) \; | \; [x:M] N \; | \; * $
+</p>
+
+$ [x:M] N $ is called a product, and $ * $ is the universe of all types, but is not a
+type itself. For an explanation of why, see the [Bonus section](#bonus).
+*Contexts* (denoted by $ \Gamma, \Delta $) are products over $*$ and have the
+form $[x_1:M_1] ... [x_n:M_n] *$. The inference system defines two judgements:
+
+1. $ \Gamma \vdash \Delta $ means $ \Delta $ is a valid context in the valid
+   context $ \Gamma $.
+2. $ \Gamma \vdash M : P $ means $ M $ is a well-typed term of type $ P $ in
+   the valid context $ \Gamma $
+
+In the inference system, a type $ P $ will either be a context or have the 
+property that $ \Gamma \vdash P : * $. Also note that $ [N/x]Q $ is the notation
+for substituting $N$ for $x$ in $Q$. The inference rules are below.
+
+<p align="center">
+<img src="fig1.png" alt="alt_text" title="image_tooltip" style="zoom:25%;" />
+</p>
+
+### Design
+
+When I started this project, I was writing my implementation based off of
+some notes written by Professor Clarkson, and this was the syntax of CoC that he used.
+
+<p align="center">
+    $ t := x \; | \; \lambda x:t_1 \rightarrow t_2 \; | \; (t_1 \; t_2) \; | \; \forall x:t_1, t_2 \; | \; \text{Type} $
+</p>
+
+I will be using this syntax throughout the rest of the blog post to go over examples,
+but the inference rules remain the same as before.
+
+I wanted the ability to write proofs in my programs, so I designed the AST with
+this in mind:
+```
+type t = 
+  | Id of id
+  | Fun of id * t * t
+  | App of t * t
+  | Forall of id * t * t
+  | Type
+
+and theorem =
+{
+  id: id;
+  theorem: t;
+  proof: t;
+}
+
+and prog = 
+| Let of id * t * prog
+| Theorem of theorem * prog
+| Expr of t
+```
+
+Here is an example of a valid program
+```
+let Nat = forall N:Type, (forall _:N, (forall _:(forall _:N, N), N)) in
+let zero = lambda N:Type -> (lambda x:N -> (lambda _:(forall _:N, N) -> x)) in
+let succ = lambda nat:Nat -> (lambda N:Type -> (lambda x:N -> 
+  (lambda f:(forall _:N, N) -> (f (((nat N) x) f))))) in
+(succ (succ zero))
+(* output: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (f x))))) *)
+```
+and this is an example of a valid (unrelated) proof.
+```
+Theorem p_implies_p: forall P:Type, forall _:P, P.
+Proof. lambda P:Type -> lambda x:P -> x.
+```
+
+### Implementation
+
+The main implementation task was implementing a typechecker. To typecheck
+a term, we first check whether that term is a context. If it is, we typecheck
+the context to ensure that it is well-typed. Otherwise, the term is typechecked
+according to the typing rules above.
```

Comment:
Hmm… maybe relating this explicitly to which judgments you're referring to in the formalism could help? In particular, I supposed I would have guessed that "we first check whether that term is a context" is _also_ specified "according to the typing rules above", namely this rule:
<img width="97" alt="Screen Shot 2022-05-24 at 4 12 27 PM" src="https://user-images.githubusercontent.com/188033/170123674-0312405b-d17a-4319-85bf-2cecfae7c586.png">